Item drop compatibility

danielzgtg opened the pull request:
Fix machines disappearing after mining/wrenching.
Fix RFTools builder losing machines.
Simplify and use methods in Block instead of calculating drops.
Make drops behave the same as mined blocks.

thiakil commented on src/main/java/mekanism/common/block/BlockBasic.java:
```diff
@@ -449,7 +447,7 @@ else if(tile instanceof TileEntityBin)
 						if(entityplayer.isSneaking())
 						{
-							dismantleBlock(state, world, pos, false);
+							dismantleBlock(state, world, pos);
```
why is there a new `dismantleBlock` method?

danielzgtg replied:
We don't need neither the ```returnBlock``` boolean anymore nor reinventing the wheel while dropping blocks with the other changes.

thiakil commented on src/main/java/mekanism/common/block/BlockBasic.java:
```diff
@@ -859,48 +862,28 @@ else if(type == BasicBlockType.INDUCTION_PROVIDER)
-			double motionZ = (world.rand.nextFloat() * motion) + (1.0F - motion) * 0.5D;
-
-			EntityItem entityItem = new EntityItem(world, pos.getX() + motionX, pos.getY() + motionY, pos.getZ() + motionZ, itemStack);
+	@Override
```
this is standard behaviour, why is it here?

danielzgtg replied:
This was used to still add stats, but not drop the block because the tile entity is missing by now.
However, I found a better solution...

thiakil commented on src/main/java/mekanism/common/block/BlockBasic.java:
```diff
@@ -859,48 +862,28 @@ else if(type == BasicBlockType.INDUCTION_PROVIDER)
 
-			world.spawnEntity(entityItem);
+	@Override
+	public void onBlockHarvested(World world, BlockPos pos, IBlockState state, EntityPlayer player) {
```
again, standard behaviour

danielzgtg replied:
The block is dropped here instead of ```harvestBlock```.

thiakil replied:
javadoc states `Called regardless of if the player's tool can actually collect this block`

danielzgtg replied:
Removed

thiakil commented on src/main/java/mekanism/common/block/BlockBounding.java:
```diff
@@ -81,14 +81,27 @@ public ItemStack getPickBlock(IBlockState state, RayTraceResult target, World wo
 	}
 
 	@Override
-	public boolean removedByPlayer(IBlockState state, World world, BlockPos pos, EntityPlayer player, boolean willHarvest)
```
This will cause bounding boxes being broken to not break their parent.

danielzgtg replied:
Oops. Fixed with new ```removedMainBlock``` method.

thiakil commented on src/main/java/mekanism/common/block/BlockBounding.java:
```diff
@@ -81,14 +81,27 @@ public ItemStack getPickBlock(IBlockState state, RayTraceResult target, World wo
+	}
+
+	@Override
+	public void harvestBlock(World world, EntityPlayer player, BlockPos pos, IBlockState state, TileEntity te, ItemStack stack) {
```
Standard behaviour copy paste

thiakil commented on src/main/java/mekanism/common/block/BlockBounding.java:
```diff
@@ -115,18 +128,6 @@ public float getPlayerRelativeBlockHardness(IBlockState state, EntityPlayer play
 		}
 	}
 
-	@Override
```
Why

danielzgtg replied:
This is not needed because ```getDrops``` is overriden.

thiakil commented on src/main/java/mekanism/common/block/BlockEnergyCube.java:
```diff
@@ -284,35 +255,25 @@ public EnumBlockRenderType getRenderType(IBlockState state)
+		if (tileEntity == null)
 		{
-			itemStack.setTagCompound(new NBTTagCompound());
+			throw new NullPointerException();
```
You must give it a message if you insist on throwing. Though I don't see much point.

danielzgtg replied:
Removed as you asked.

thiakil commented on src/main/java/mekanism/common/block/BlockEnergyCube.java:
```diff
@@ -284,35 +255,25 @@ public EnumBlockRenderType getRenderType(IBlockState state)
+			throw new NullPointerException();
 		}
-		
-		if(tileEntity instanceof ISecurityTile)
```
why are you removing instance checks?

danielzgtg replied:
There was a warning that ```TileEntityEnergyCube``` is always ```instanceof ISecurityTile```.

thiakil commented on src/main/java/mekanism/common/block/BlockTransmitter.java:
```diff
@@ -231,49 +230,54 @@ public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, En
 		{
-			itemStack = new ItemStack(MekanismBlocks.Transmitter, 1, tileEntity.getTransmitterType().ordinal());
-			
-			if(!itemStack.hasTagCompound())
```
This will wipe any data a capability may have added.

danielzgtg replied:
The constructor ```ItemStack(Block, int, int)``` (calls ```ItemStack(Item, int, int)```) is being used, so the ```NBTTagCompound``` should always be null.

thiakil replied:
It does not matter what constructor you use.

danielzgtg replied:
I must have overlooked ```forgeInit```

danielzgtg commented:
Summary of what is being done:
Previously, Mekanism used ```removedByPlayer``` to drop the block.
This was a problem because RFTools' builder uses ```getDrops```, which was not returning any items because ```quantityDropped``` was returning 0.

The original pull request made ```getDrops``` return the proper amount of items. It did not work until drops were moved ```onBlockHarvested``` since the tile entity was removed before ``harvestDrops```.

The second set of changes did not copy and paste vanilla code and used ```removedByPlayer``` to delay removal of the tile entity. Unfortunately, I just found out the new solution caused inventory loss, so I reverted it, and I am still working on it.

thiakil commented:
`onBlockHarvested` is called by `removedByPlayer` (and only by it) regardless of whether the player has the correct tool or not. The `willHarvest` flag needs to be checked.

danielzgtg commented:
# New changes #
1. Removed duplicated vanilla code in normal ```harvestBlock```
2. Added new edited vanilla code to ```harvestBlock``` for ```BlockContainer``` to fix inventory loss caused by change 1.
3. Fixed boundary blocks broken by original pull request.
4. Extracted ```dismantleBlock``` into utils.
5. Restored ```hasTagCompound``` code.

Now waiting on McJty/RFTools#1688. This pull request will cause a DigiMiner duplication glitch with RFTools builder otherwise.

# Some testing #
Test environment: Singleplayer, survival & creative, normal diamond pickaxe, silk diamond pickaxe, crescent hammer (thermal expansion), RFTools builder.

Harvesting these seems to work:
1. Digiminer (empty)
2. Digiminer with 1 grass
2. Osmium & obsidian storage blocks
3. Harvesting enrichment chamber with obsidian storage block
4. Harvesting enrichment chamber with the item from test 4

thiakil commented on src/main/java/mekanism/common/block/BlockBounding.java:
```diff
@@ -76,20 +86,82 @@ public ItemStack getPickBlock(IBlockState state, RayTraceResult target, World wo
 			IBlockState state1 = world.getBlockState(tileEntity.mainPos);
 			return state1.getBlock().getPickBlock(state1, target, world, tileEntity.mainPos, player);
 		} catch(Exception e) {
+			e.printStackTrace();
```
no

thiakil commented on src/main/java/mekanism/common/block/BlockBounding.java:
```diff
@@ -76,20 +86,82 @@ public ItemStack getPickBlock(IBlockState state, RayTraceResult target, World wo
+			state1.getBlock().getDrops(drops, world, mainPos, state1, fortune);
 		} catch(Exception e) {
-			return false;
+			e.printStackTrace();
```
log it properly or rewrite it so it doesn't need the try catch (the better option)

danielzgtg replied:
What are we trying to catch anyway? If it's just a ```NullPointerException```, this can be simply replaced with a null check.

thiakil replied:
That's my best guess as to what it was for, so yeah, just null check it

thiakil commented on src/main/java/mekanism/common/block/BlockBounding.java:
```diff
@@ -76,20 +86,82 @@ public ItemStack getPickBlock(IBlockState state, RayTraceResult target, World wo
+
+	private void removeMainBlock(World world, BlockPos pos) {
+		TileEntityBoundingBlock tileEntity = (TileEntityBoundingBlock)world.getTileEntity(pos);
+		world.setBlockToAir(tileEntity.mainPos);
```
Setting the main block to air doesnt seem right to me...

danielzgtg replied:
It works fine, because we drop the items in ```getDrops``` on behalf of the main block.

The main block shouldn't mind not being told about being removed, because all the ```removedByPlayer``` and ```harvestBlock``` hooks were just added.
All of the other bounding blocks are removed with block updates.

thiakil replied:
> It works fine, because we drop the items in getDrops on behalf of the main block.

That is bad practice. The main block should be handling its own breakage.

danielzgtg replied:
It's not the best but if the main block were to handle its own harvesting, it would drop itself as an itemstack into the world.
That means I can't pack up my base with RFTools builder anymore (or probably even Buildcraft quarry), which is the purpose of this pull request.
Another way is to add a method to the main block, to ask it to give us the items, but isn't that just ```getDrops```?

danielzgtg replied:
Doing a ```getDrops``` and ```setBlockToAir``` seems to be standard, and will be done to us by other mods:
1. https://github.com/McJty/RFTools/blob/496527a1faa37ad16b60d2ea10751074c3d7d2e8/src/main/java/mcjty/rftools/blocks/builder/BuilderTileEntity.java#L1136-L1144
2. https://findusages.com/search/net.minecraft.block.Block/getDrops$4
3. https://findusages.com/search/net.minecraft.block.Block/getDrops$5

thiakil commented on src/main/java/mekanism/common/block/BlockTransmitter.java:
```diff
@@ -231,49 +230,84 @@ public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, En
-			
-			if(!itemStack.hasTagCompound())
+			ItemStack itemStack = new ItemStack(MekanismBlocks.Transmitter, 1,
+					tileEntity.getTransmitterType().ordinal());
```
this line does not need to be split

danielzgtg replied:
Removed most IDE and non-essential changes.

thiakil commented on src/main/java/mekanism/common/block/BlockTransmitter.java:
```diff
@@ -231,49 +230,84 @@ public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, En
-			if(!itemStack.hasTagCompound())
+			ItemStack itemStack = new ItemStack(MekanismBlocks.Transmitter, 1,
+					tileEntity.getTransmitterType().ordinal());
+			if (!itemStack.hasTagCompound())
```
needless format change

josephcsible commented:
> Now waiting on McJty/RFTools#1688. This pull request will cause a DigiMiner duplication glitch with RFTools builder otherwise.

Done in RFTools 7.32.

josephcsible commented:
Why was this closed?

dizzyd commented:
Hi @josephcsible - this PR got closed as we are in the process of merging Mekanica (a short-lived fork) back into Mekanism. As you can see, there are now some conflicts, so if you're still interested in having this merged, we'll need to get those resolved. Thanks! :)

pupnewfster commented:
Closing this as I ported the changes to be able to merge cleanly in #5384 once we have a chance to further review and test this.
